File: packages/schema/src/localize.ts
/**
 * The projection: `Card` (all locales) -> `LocalizedCard` (one locale).
 *
 * **Port of `src/holo_schema/localize.py`.** That file is the reference implementation
 * and carries the full reasoning for each merge rule; this is the copy the Worker runs,
 * because D8 stores translations as JSON and the flattening therefore happens on read,
 * inside the Worker, in TypeScript.
 *
 * Two implementations, one spec. `tests/localize.test.ts` asserts this reproduces the
 * golden files that the Python side generates. If they diverge, `make check` fails.
 *
 * When editing this file, edit `localize.py` in the same commit.
 */

import type { Card } from "../dist/card.d.ts";
import type { LocalizedCard, LocalizedArt } from "../dist/localized-card.d.ts";
import { SOURCE_LOCALE, type Locale } from "../dist/enums.ts";

/**
 * Project a canonical card into the API's single-locale shape.
 *
 * @param card - the canonical card, carrying all locales.
 * @param locale - the language to project into. Falls back to the source locale if the
 *   card has no translation for it.
 * @returns the card as the API serves it, with translation fields flattened.
 * @throws if the card has neither the requested locale nor the source locale.
 */
export function localize(card: Card, locale: Locale): LocalizedCard {
  const translations = card.translations as Record<string, Card["translations"][string]>;

  let translation = translations[locale];
  let resolved: Locale = locale;

  if (translation === undefined) {
    translation = translations[SOURCE_LOCALE];
    resolved = SOURCE_LOCALE;
  }

  if (translation === undefined) {
    throw new Error(
      `card ${card.id} has no translation for '${locale}' and no '${SOURCE_LOCALE}' fallback`,
    );
  }

  // --- Arts: zip the language-independent half with the translated half by index ---
  //
  // The translated list may be shorter (hSD03-009 and hSD04-009 have 2 arts but 0 `en`
  // translations). Emit the art anyway, unnamed — dropping it would misreport the card.
  const baseArts = card.arts ?? [];
  const translatedArts = translation.arts ?? [];
  const arts: LocalizedArt[] = baseArts.map((base, index) => {
    const translated = translatedArts[index];
    return {
      cost_count: base.cost_count,
      cost_types: base.cost_types ?? [],
      ...(base.damage !== undefined && { damage: base.damage }),
      ...(base.is_plus !== undefined && { is_plus: base.is_plus }),
      ...(base.special_targets !== undefined && { special_targets: base.special_targets }),
      ...(base.special_values !== undefined && { special_values: base.special_values }),
      ...(translated?.name !== undefined && { name: translated.name }),
      ...(translated?.effect !== undefined && { effect: translated.effect }),
    };
  });

  // --- Keyword: type code from the card, display text from the translation ---
  const keyword = card.keyword
    ? {
        type: card.keyword.type,
        type_code: card.keyword.type_code,
        ...(translation.keyword?.name !== undefined && { name: translation.keyword.name }),
        ...(translation.keyword?.effect !== undefined && {
          effect: translation.keyword.effect,
        }),
      }
    : undefined;

  // --- Oshi skills: timing code from the card, display text from the translation ---
  const oshiSkill = card.oshi_skill
    ? {
        timing_code: card.oshi_skill.timing_code,
        ...(translation.oshi_skill?.name !== undefined && {
          name: translation.oshi_skill.name,
        }),
        ...(translation.oshi_skill?.effect !== undefined && {
          effect: translation.oshi_skill.effect,
        }),
        ...(translation.oshi_skill?.timing !== undefined && {
          timing: translation.oshi_skill.timing,
        }),
      }
    : undefined;

  const spOshiSkill = card.sp_oshi_skill
    ? {
        timing_code: card.sp_oshi_skill.timing_code,
        ...(translation.sp_oshi_skill?.name !== undefined && {
          name: translation.sp_oshi_skill.name,
        }),
        ...(translation.sp_oshi_skill?.effect !== undefined && {
          effect: translation.sp_oshi_skill.effect,
        }),
        ...(translation.sp_oshi_skill?.timing !== undefined && {
          timing: translation.sp_oshi_skill.timing,
        }),
      }
    : undefined;

  return {
    id: card.id,
    card_number: card.card_number,
    locale: resolved,
    card_type_code: card.card_type_code,
    rarity_code: card.rarity_code,
    color_codes: card.color_codes ?? [],
    ...(card.bloom_level_code !== undefined && { bloom_level_code: card.bloom_level_code }),
    image_key: card.image_key,
    ...(card.hp !== undefined && { hp: card.hp }),
    ...(card.life !== undefined && { life: card.life }),
    ...(card.baton_touch_count !== undefined && {
      baton_touch_count: card.baton_touch_count,
    }),
    baton_touch_types: card.baton_touch_types ?? [],
    ...(card.illustrator !== undefined && { illustrator: card.illustrator }),
    card_sets: card.card_sets,
    name: translation.name,
    tags: translation.tags ?? [],
    ...(translation.ability_text !== undefined && {
      ability_text: translation.ability_text,
    }),
    ...(translation.extra !== undefined && { extra: translation.extra }),
    arts,
    ...(keyword !== undefined && { keyword }),
    ...(oshiSkill !== undefined && { oshi_skill: oshiSkill }),
    ...(spOshiSkill !== undefined && { sp_oshi_skill: spOshiSkill }),
    qa_items: translation.qa_items ?? [],
  } as LocalizedCard;
}

/**
 * Compose a card's CDN URL from its key (decision D9).
 *
 * The one place the key -> URL mapping lives on the TypeScript side. v1 copy-pasted a
 * `getImagePath()` helper verbatim across three components; per the architecture review
 * that duplication is retired by having exactly one composer.
 *
 * @param imageKey - e.g. "default/hBP01-028_C_02"
 * @param baseUrl - CDN origin, e.g. "https://img.hololive-ocg-wiki.tskrlabs.com"
 * @param extension - image format. WebP only in production per D9.
 */
export function cardImage(
  imageKey: string,
  baseUrl: string,
  extension = "webp",
): string {
  return `${baseUrl.replace(/\/$/, "")}/${imageKey}.${extension}`;
}

File: packages/schema/src/holo_schema/localize.py
"""The projection: `Card` (all locales) -> `LocalizedCard` (one locale).

**This is the reference implementation.** A TypeScript port lives in
`packages/schema/src/localize.ts` because the Worker must run this projection at
request time (D8 stores translations as JSON, so the flattening happens on read, and
the Worker is TypeScript).

Two implementations, one spec. `tests/test_localize.py` writes golden files from this
one; `tests/localize.test.ts` asserts the TypeScript reproduces them exactly. If the
two ever disagree, `make check` fails.

The merge rules, all of which the data forced:

1. **Locale fallback.** A missing locale falls back to the source locale (`ja`) rather
   than returning nothing. All 7 locales are present on all cards today, so this never
   fires — but an 8th locale is added translation-by-translation, and a half-translated
   card should render in Japanese rather than 404.

2. **Arts pair by index, tolerating a short list.** hSD03-009 and hSD04-009 had 2 arts
   but 0 `en` translations. The art is emitted with costs and damage and no name, since
   dropping it would misreport what the card does. F-004 resolved itself and no real
   card has a short list any more — a census over all 2,463 finds zero mismatches in any
   locale — so the **synthetic fixture `9000001` is the only thing covering this
   branch**, in both implementations. Deleting it silently untests a rule that ships.
   See SYNTHETIC_CARD in `scripts/build_fixtures.py`.

3. **Excess translated arts are dropped.** The inverse case does not occur in today's
   data. An art with a name but no cost is not a valid art, so there is nothing
   meaningful to emit.

4. **Absent list fields become empty lists.** `Card` distinguishes "no tags" (absent)
   from "empty tags"; the API does not, because every consumer does `v-if="tags.length"`
   and an optional array just moves the null check into the frontend.
"""

from .card import Card
from .enums import SOURCE_LOCALE, Locale
from .localized import (
    LocalizedArt,
    LocalizedCard,
    LocalizedKeyword,
    LocalizedOshiSkill,
)


def localize(card: Card, locale: Locale) -> LocalizedCard:
    """Project a canonical `Card` into the API's single-locale shape.

    Args:
        card: the canonical card, carrying all locales.
        locale: the language to project into. Falls back to the source locale if the
            card has no translation for it.

    Returns:
        The card as the API serves it, with translation fields flattened to the top
        level and nested data merged.

    Raises:
        ValueError: if the card has neither the requested locale nor the source locale.
            `Card`'s validator makes this unreachable for validated cards.
    """
    translation = card.translations.get(locale)
    resolved: Locale = locale

    if translation is None:
        translation = card.translations.get(SOURCE_LOCALE)
        resolved = SOURCE_LOCALE

    if translation is None:
        raise ValueError(
            f"card {card.id} has no translation for '{locale}' and no "
            f"'{SOURCE_LOCALE}' fallback"
        )

    # --- Arts: zip the language-independent half with the translated half by index ---
    base_arts = card.arts or []
    translated_arts = translation.arts or []
    arts: list[LocalizedArt] = []
    for index, base in enumerate(base_arts):
        translated = translated_arts[index] if index < len(translated_arts) else None
        arts.append(
            LocalizedArt(
                cost_types=base.cost_types,
                damage=base.damage,
                is_plus=base.is_plus,
                special_targets=base.special_targets,
                special_values=base.special_values,
                name=translated.name if translated else None,
                effect=translated.effect if translated else None,
            )
        )

    # --- Keyword: type code from the card, display text from the translation ---
    keyword = None
    if card.keyword is not None:
        keyword = LocalizedKeyword(
            type=card.keyword.type,
            type_code=card.keyword.type_code,
            name=translation.keyword.name if translation.keyword else None,
            effect=translation.keyword.effect if translation.keyword else None,
        )

    # --- Oshi skills: timing code from the card, display text from the translation ---
    oshi_skill = None
    if card.oshi_skill is not None:
        oshi_skill = LocalizedOshiSkill(
            timing_code=card.oshi_skill.timing_code,
            name=translation.oshi_skill.name if translation.oshi_skill else None,
            effect=translation.oshi_skill.effect if translation.oshi_skill else None,
            timing=translation.oshi_skill.timing if translation.oshi_skill else None,
        )

    sp_oshi_skill = None
    if card.sp_oshi_skill is not None:
        sp_oshi_skill = LocalizedOshiSkill(
            timing_code=card.sp_oshi_skill.timing_code,
            name=translation.sp_oshi_skill.name if translation.sp_oshi_skill else None,
            effect=(
                translation.sp_oshi_skill.effect if translation.sp_oshi_skill else None
            ),
            timing=translation.sp_oshi_skill.timing if translation.sp_oshi_skill else None,
        )

    return LocalizedCard(
        id=card.id,
        card_number=card.card_number,
        locale=resolved,
        card_type_code=card.card_type_code,
        rarity_code=card.rarity_code,
        color_codes=card.color_codes or [],
        bloom_level_code=card.bloom_level_code,
        image_key=card.image_key,
        hp=card.hp,
        life=card.life,
        baton_touch_count=card.baton_touch_count,
        baton_touch_types=card.baton_touch_types or [],
        illustrator=card.illustrator,
        card_sets=card.card_sets,
        name=translation.name,
        tags=translation.tags or [],
        ability_text=translation.ability_text,
        extra=translation.extra,
        arts=arts,
        keyword=keyword,
        oshi_skill=oshi_skill,
        sp_oshi_skill=sp_oshi_skill,
        qa_items=translation.qa_items or [],
    )
